### DISCUSSION TITLE
[HW2] Question about problem 3.1

### DISCUSSION THREAD

**Discussion by dinhkt** [seominjoon/kaist-ai605#104] [Q&A] [answered]
Hi Professor,
In problem 3.1, what is Pytorch's tensor native for operation for dense search?

**Comment by seominjoon** [comment, marked as answer]
You are computing inner product between vectors for obtaining the scores, and there are several ways. One way is performing matrix multiplication with (native) PyTorch. Another way can be computing via pure Python (e.g. looping through each vector as a Python list), which I don't recommend. Another method would be using Faiss, which is Problem 3.3.

**Reply by emjay73** [reply]
How about using NumPy matmul? (or @)?

**Reply by seominjoon** [reply]
numpy's matmul won't be good either. SciPy's sparse matmul could work though.

**Reply by emjay73** [reply]
Even for dense search, should we consider sparsity?

**Reply by seominjoon** [reply]
Oh my bad, I was confused with another question. For dense, using numpy is fine too.
